### ISSUE TITLE
Citations have month and day next to year

### ISSUE THREAD

**Issue by gnpan** [retorquere/zotero-better-bibtex#868]
- [x] I am on the latest release of BBT and Zotero
- [x] I am posting a single question, bug or feature request
- [x] I have included a descriptive subject of the problem
- [x] I am available for follow-up questions and testing

Exporting collection with BibLaTeX
Report ID: 74AZDIFM

Using pandoc, cell.csl, and standard latex template. Resulting citations and bibliography include year, month and day, instead of just year, as was happening before. Old bib file with the same references exported with Zotero 4 works OK. Comparison between old and new (v.5) bib file contents shows the following difference:
v.4: date = {1996-11-15}
v.5: year = {1996 Nov 15}

Thanks in advance for any suggestions.

**Comment by blip-bloop** [comment]
:robot: this is your friendly neighborhood build bot announcing [test build 5281](https://github.com/retorquere/zotero-better-bibtex/releases/download/builds/zotero-better-bibtex-5.0.66.5281.issue-868.xpi) ("fixes #868").

**Comment by retorquere** [comment]
Can you try 5281?

**Comment by gnpan** [comment]
Still the same with 5281, but deleting and re-importing the offending references seems to correct the issue - all new additions to the library also seem to work OK. 
Many thanks for the quick response.

**Comment by retorquere** [comment]
Hmm, that means the cache isn't dropped after the install. I'll see if I can find the cause if that.

**Comment by retorquere** [comment]
Are you open to helping with the cache problem? I can't yet reproduce it. To see if the cache issue is reproducible:

1. Install BBT 5.0.65
2. import the reference with `1996 Nov 15`
3. Install 5281
4. Send a BBT error report

Can you also right-click a reference of which you say "old references that still do not export OK have only month and year in the date field" and send a error report? I need to see the exact contents of those date fields.

**Comment by blip-bloop** [comment]
:robot: this is your friendly neighborhood build bot announcing [test build 5286](https://github.com/retorquere/zotero-better-bibtex/releases/download/builds/zotero-better-bibtex-5.0.66.5286.issue-868.xpi) ("be chatty about cache clear activity").

**Comment by gnpan** [comment]
Error report for old reference that still does not export correctly: ECD9X4XE

**Comment by gnpan** [comment]
PB6972BW
And this is the report for
1. Install BBT 5.0.65
2. import the reference with 1996 Nov 15
3. Install 5281
4. Send a BBT error report

**Comment by retorquere** [comment]
Could you do the same 4-steps with 5286? It won't fix it but it has extra logging.

**Comment by gnpan** [comment]
Here it is with 5286: 49ZGVL46

**Comment by retorquere** [comment]
And 49ZGVL46 is right after restarting after the install of 5286?

**Comment by blip-bloop** [comment]
:robot: this is your friendly neighborhood build bot announcing [test build 5293](https://github.com/retorquere/zotero-better-bibtex/releases/download/builds/zotero-better-bibtex-5.0.69.5293.issue-868.xpi) ("test case").

**Comment by gnpan** [comment]
Yes, that's right.

**Comment by gnpan** [comment]
5293 fixes it - no more year={...
Many thanks!

**Comment by retorquere** [comment]
Sure, but the caching issue is a problem. Unless 5293 also magically (because I've only added logging) resolves the cache drop problem for you.

**Comment by gnpan** [comment]
Can't tell what it is, but I reinstalled 5286 and the problem came back. Then I installed 5293 again and all refs exported fine.

**Comment by retorquere** [comment]
Wait - if you install 5286 the problem comes back? That can't be a cache issue.

**Comment by gnpan** [comment]
Just tried it again, yes the issue comes back with 5286 for references with month and year. As I said before, 5281 fixed it for references with day, month and year. 5293 fixes it for all references. Let me know if you want me to try something else.

**Comment by retorquere** [comment]
Just so I understand correctly, you're not re-importing references or changing them,just exporting the same unchanged reference from your library with these two versions?

**Comment by gnpan** [comment]
That's right, these are the original references, imported some time last year.

**Comment by retorquere** [comment]
OK, help me think here.

## what's supposed to happen

On shutdown, BBT writes the cache to disk, noting in it the version of BBT that wrote it. On startup, BBT reads the cache, but erases the contents if the version number stored in the cache does not match the running version.

## what we see happening

5286 does not seem to drop the cache after first run. In 49ZGVL46, I see `CACHE: retaining cache Better BibLaTeX because stored BetterBibTeX is 5.0.66.5286.issue-868 (current: 5.0.66.5286.issue-868)`, which I would have expected to see at a 2nd or later start of 5286, but if you did the 4-step, it should have said `CACHE: dropping cache Better BibTeX because BetterBibTeX went from 5.0.65 to 5.0.66.5286.issue-868`. I don't yet know why this happens.

Wait. Did you restart Zotero after each install?

## the second, even weirder behavior

Just so I have it clear, this is what you have done (and nothing else):

1. installed 5293, restarted, and exported. Exports are OK.
2. installed 5286, restarted, and exported. Exports are *not* OK.
1. installed 5293, restarted, and exported. Exports are OK.
2. installed 5286, restarted, and exported. Exports are *not* OK.

The problem I have with this is that even if 5286 for some reason did not drop the cache, 5293 obviously does (or you would not have seen the fixed exports). So why do the broken exports come back in step 2. and 4.? It makes no sense. These builds have the fixed exporter so the only remaining possible cause is that they're still in the cache, and a re-import and export of a new reference shows that it has the fixed export. So where is the unfixed data coming from, if at least step 1. and 3. dropped the cache?

**Comment by gnpan** [comment]
Yes, I did restart Zotero after each install. The cache issue is beyond me. I just tried again: moved 5 random refs in a new collection and went back to 5.0.65. The exported file has 3 refs with year=... instead of date=... Then I installed 5286: One ref was fixed, the other two remained with "year". Then installed 5293, every ref is OK with "date". 
Would deleting Zotero and all its files, caches, etc and re-installing it be of any use?

**Comment by retorquere** [comment]
Oh wait, no, that's all explainable. The fix for `1996 Nov 15` was in both, but `1996 Nov` required a 2nd fix and that wasn't in 5286. So perhaps this wasn't a cache issue at all -- I thought all this happened on `1996 Nov 15`.

**Comment by retorquere** [comment]
(date parsing is very messy and there's lots of special cases to deal with. The date parser in the 5.0 series is more robust but it didn't carry along all the edge case handling)

**Comment by gnpan** [comment]
OK, that makes sense. The 1996 reference was just an example I sent you. The issue affected around 100 refs which had been imported in one go sometime last year - maybe there was something funny with the way dates were imported. It's easy to re-import them all just in case, although 5293 deals fine with all of them. 
Again, many thanks for your great help.

**Comment by retorquere** [comment]
If I've done my job well, no re-importing should be necessary. The cache should just be dropped, and new exports should get the fixed date fields.

**Comment by retorquere** [comment]
There can't be anything funny about how the dates were imported BTW. Zotero imports dates as free-form text and only parses them to structured dates as needed (o.a. for BibTeX exports, but also other places). BBT has it's own "date-ish text"→"structured data" parser which handles non-structured dates differently than the Zotero date parser.

**Comment by gnpan** [comment]
Good to know. Everything works OK so far without re-importing the references, so I guess you have done your job well ;)
By the way, this was the fastest and most efficient support I ever had on any software!
